server/impl/nonce/nonce.go

```diff
@@ -0,0 +1,76 @@
+package nonce
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+)
+
+const (
+	stateLen       = 16
+	stateExpiry    = 30 * time.Minute
+	clearThreshold = 128
+)
+
+// Generator creates randomly generated 16-byte ASCII nonces. The nonce only has
+// Base32 characters, which are a subset of all alphanumeric characters.
+// Each nonce is only valid for 30 minutes; beyond that, and VerifyState will
+// return false.
+type Generator struct {
+	rand        *rand.Rand
+	m           map[string]time.Time
```

Comment:
Ideally we should store the temporary state in a DB (à la Redis) that would persist across server restarts as well as multiple servers, but we don't have either scenario here. So this should do for now.